scipy/special/cephes/igami.c

```diff
@@ -1,180 +1,342 @@
+{
+    /*
+     * Computation of the Incomplete Gamma Function Ratios and their Inverse
+     * ARMIDO R. DIDONATO and ALFRED H. MORRIS, JR.
+     * ACM Transactions on Mathematical Software, Vol. 12, No. 4,
+     * December 1986, Pages 377-393.
+     *
+     * See equation 32.
+     */
+    double s, t;
+    double a[4] = {0.213623493715853, 4.28342155967104,
+		   11.6616720288968, 3.31125922108741};
+    double b[5] = {0.3611708101884203e-1, 1.27364489782223,
+		   6.40691597760039, 6.61053765625462, 1};
 
-    if (npy_isnan(a) || npy_isnan(y0)) {
-        return NPY_NAN;
+    if (p < 0.5) {
+        t = sqrt(-2 * log(p));
     }
-    if ((y0 < 0.0) || (y0 > 1.0) || (a <= 0)) {
-	mtherr("igami", DOMAIN);
-	return NPY_NAN;
+    else {
+        t = sqrt(-2 * log(q));
     }
-    if (y0 == 0.0) {
-	return NPY_INFINITY;
+    s = t - polevl(t, a, 3) / polevl(t, b, 4);
+    if(p < 0.5)
+        s = -s;
+    return s;
+}
+
+
+double didonato_SN(double a, double x, unsigned N, double tolerance)
+{
+    /*
+     * Computation of the Incomplete Gamma Function Ratios and their Inverse
+     * ARMIDO R. DIDONATO and ALFRED H. MORRIS, JR.
+     * ACM Transactions on Mathematical Software, Vol. 12, No. 4,
+     * December 1986, Pages 377-393.
+     *
+     * See equation 34.
+     */
+    double sum = 1.0;
+
+    if (N >= 1) {
+	unsigned i;
+        double partial = x / (a + 1);
+
+        sum += partial;
+        for(i = 2; i <= N; ++i) {
+            partial *= x / (a + i);
+            sum += partial;
+            if(partial < tolerance) {
+                break;
+            }
+        }
     }
-    if (y0 == 1.0) {
-	return 0.0;
+    return sum;
+}
+
+
+double find_inverse_gamma(double a, double p, double q)
+{
+    /*
+     * In order to understand what's going on here, you will
+     * need to refer to:
+     *
+     * Computation of the Incomplete Gamma Function Ratios and their Inverse
+     * ARMIDO R. DIDONATO and ALFRED H. MORRIS, JR.
+     * ACM Transactions on Mathematical Software, Vol. 12, No. 4,
+     * December 1986, Pages 377-393.
+     */
+    double result;
+
+    if (a == 1) {
+        if (q > 0.9) {
+            result = -log1p(-p);
+        }
+        else {
+            result = -log(q);
+        }
+    }
+    else if (a < 1) {
+        double g = Gamma(a);
+        double b = q * g;
+
+        if ((b > 0.6) || ((b >= 0.45) && (a >= 0.3))) {
+            /* DiDonato & Morris Eq 21:
+             *
+             * There is a slight variation from DiDonato and Morris here:
+             * the first form given here is unstable when p is close to 1,
+             * making it impossible to compute the inverse of Q(a,x) for small
+             * q. Fortunately the second form works perfectly well in this case.
+             */
+            double u;
+            if((b * q > 1e-8) && (q > 1e-5)) {
+                u = pow(p * g * a, 1 / a);
+            }
+            else {
+                u = exp((-q / a) - NPY_EULER);
```

Comment:
No test coverage for this branch, according to `runtests.py --gcov -s special; runtests.py --lcov-html`